ports/stm32f4/common-hal/busio/I2C.c

```diff
@@ -32,52 +32,129 @@
 
 #include "shared-bindings/microcontroller/__init__.h"
 #include "supervisor/shared/translate.h"
+#include "common-hal/microcontroller/Pin.h"
 
-I2C_HandleTypeDef hi2c2;
+void i2c_reset(void) {
+    //TODO: implement something better than eratta workaround. 
+}
 
 void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
         const mcu_pin_obj_t* scl, const mcu_pin_obj_t* sda, uint32_t frequency, uint32_t timeout) {
-    //TODO: Rework this entire section to use LL so we can properly assign pins
-    //      This will also be bundled with MSP removal
-    hi2c2.Instance = I2C1;
-    hi2c2.Init.ClockSpeed = 100000;
-    hi2c2.Init.DutyCycle = I2C_DUTYCYCLE_2;
-    hi2c2.Init.OwnAddress1 = 0;
-    hi2c2.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
-    hi2c2.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
-    hi2c2.Init.OwnAddress2 = 0;
-    hi2c2.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
-    hi2c2.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
-    if(HAL_I2C_Init(&hi2c2) != HAL_OK) {
-        mp_raise_RuntimeError(translate("I2C Init Error"));
-    } else {
-        mp_printf(&mp_plat_print, "I2C INIT OK");
+
+    //match pins to I2C objects
+    I2C_TypeDef * I2Cx;
+
+    uint8_t sda_len = sizeof(mcu_i2c_sda_list)/sizeof(*mcu_i2c_sda_list);
+    uint8_t scl_len = sizeof(mcu_i2c_scl_list)/sizeof(*mcu_i2c_scl_list);
+    for(uint i=0; i<sda_len;i++) {
+        if (mcu_i2c_sda_list[i]->pin == sda) {
+            for(uint j=0; j<scl_len;j++) {
+                if ((mcu_i2c_scl_list[j]->pin == scl)
+                    && (mcu_i2c_scl_list[j]->i2c_index == mcu_i2c_sda_list[i]->i2c_index)) {
+                    self->scl = mcu_i2c_scl_list[j];
+                    self->sda = mcu_i2c_sda_list[i];
+                    break;
+                }
+            }
+        }
     }
 
-    self->sda_pin = sda->number;
-    self->scl_pin = scl->number;
+    //handle typedef selection, errors
+    if(self->sda!=NULL && self->scl!=NULL ) {
+        I2Cx = mcu_i2c_banks[self->sda->i2c_index-1];
+    } else {
+        mp_raise_RuntimeError(translate("Invalid I2C pin selection"));
+    }
 
```

Comment:
Check here to see if the pins are already in use too. You can test it by constructing the same bus twice. The second time should fail with pin in use.

Reply:
Isn't this already done at the application level? I get this behavior by default, and neither the samd or nrf ports have claimed pin checking. Or did you mean checking the I2C peripherals themselves? That's a good idea, since trying to initialize the same peripheral twice will cause IO errors on writing, but I'm not sure how to present that to the user given how we've abstracted the device away.

Reply:
Yup! You are right. The pins are checked here: https://github.com/adafruit/circuitpython/blob/master/shared-bindings/busio/I2C.c#L78

Checking the peripherals would be good too. Looks like [SAMD will do "Invalid pins"](https://github.com/adafruit/circuitpython/blob/master/ports/atmel-samd/common-hal/busio/I2C.c#L76) but something like "Hardware in use" is probably better. 

Also make sure that after you deinit you can recreate a new bus with the same pins. deinit should have freed them.

Reply:
Whoop good point on the deinit I think I forgot that